mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogReader.cs

```diff
@@ -0,0 +1,147 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace Mono.Profiler.Log {
+
+	public sealed class LogReader : IDisposable {
+
+		public static Encoding Encoding { get; } = Encoding.UTF8;
+
+		public LogStream BaseStream => (LogStream) _reader.BaseStream;
+
+		public long Position { get; private set; }
+
+		readonly BinaryReader _reader;
+
+		public LogReader (LogStream stream, bool leaveOpen)
+		{
+			_reader = new BinaryReader (stream, Encoding, leaveOpen);
+		}
+
+		public void Dispose ()
+		{
+			_reader.Dispose ();
+		}
+
+		internal byte ReadByte ()
+		{
+			var b = _reader.ReadByte ();
+
+			Position += sizeof (byte);
+
+			return b;
+		}
+
+		internal ushort ReadUInt16 ()
+		{
+			var i = _reader.ReadUInt16 ();
+
+			Position += sizeof (ushort);
+
+			return i;
+		}
+
+		internal int ReadInt32 ()
+		{
+			var i = _reader.ReadInt32 ();
+
+			Position += sizeof (int);
+
+			return i;
+		}
+
+		internal long ReadInt64 ()
+		{
+			var i = _reader.ReadInt64 ();
+
+			Position += sizeof (long);
+
+			return i;
+		}
+
+		internal ulong ReadUInt64 ()
+		{
+			var i = _reader.ReadUInt64 ();
+
+			Position += sizeof (ulong);
+
+			return i;
+		}
+
+		internal double ReadDouble ()
+		{
+			var d = _reader.ReadDouble ();
+
+			Position += sizeof (double);
+
+			return d;
+		}
+
+		internal string ReadHeaderString ()
+		{
+			var bytes = _reader.ReadBytes (ReadInt32 ());
+
+			Position += bytes.Length;
+
+			return Encoding.GetString (bytes);
+		}
+
+		internal string ReadCString ()
+		{
+			var bytes = new List<byte> ();
```

Comment:
Since every name is read via this method, maybe it's a good idea to reuse a buffer or something here, or at least pre-alloc the list to something bigger.